taichi/analysis/gather_used_atomics.cpp

```diff
@@ -5,20 +5,21 @@ TLANG_NAMESPACE_BEGIN
 
 class UsedAtomicsSearcher : public BasicStmtVisitor {
  private:
-  std::unordered_set<AtomicOpStmt *> used_atomics;
+  std::unique_ptr<std::unordered_set<AtomicOpStmt *>> used_atomics;
```

Comment:
If the goal is to avoid copying, `std::unordered_set` itself is movable and we don't need `std::unique_ptr`.

Reply:
Oh I didn't know it. Will `return searcher.used_atomics;` avoid copying in `static std::unordered_set<AtomicOpStmt *> run(IRNode *root)` (where `searcher` is a temporary variable)?

Reply:
That depends on a few features of C++: copy elision and return value optimization. You might also need `return std::move(...);`, which might or might not be helpful if you have copy elision and RVO...

Reply:
I just realized that I used `std::unique_ptr` to distinguish the uninitialized case (which should not happen) and an empty set.